src/main/java/dev/morling/onebrc/CalculateAverage_thomaswue.java

```diff
@@ -175,42 +174,25 @@ else if (((nextVal & 0x3B000000) == 0x3B000000)) {
             }
 
             // Final calculation for index into hash table.
-            int tableIndex = (((hash ^ (hash >>> 18)) & (results.length - 1)));
-            while (true) {
+            int hashAsInt = (int) (hash ^ (hash >>> 32));
+            int finalHash = (hashAsInt ^ (hashAsInt >>> 18));
+            int tableIndex = (finalHash & (results.length - 1));
+            outer: while (true) {
                 Result existingResult = results[tableIndex];
                 if (existingResult == null) {
                     newEntry(results, cities, nameAddress, number, tableIndex, nameLength);
                     break;
                 }
                 else {
                     // Check for collision.
-                    boolean result = true;
                     int i = 0;
-                    if ((long) nameLength >= 8) {
-                        if (UNSAFE.getLong(existingResult.nameAddress) != UNSAFE.getLong(nameAddress)) {
-                            result = false;
-                        }
-                        else {
-                            i += 8;
+                    for (; i < nameLength - 8; i += 8) {
+                        if (UNSAFE.getLong(existingResult.nameAddress + i) != UNSAFE.getLong(nameAddress + i)) {
+                            tableIndex = (tableIndex + 1) & (results.length - 1);
+                            continue outer;
                         }
                     }
-                    else if ((long) nameLength >= 4) {
-                        if (UNSAFE.getInt(existingResult.nameAddress) != UNSAFE.getInt(nameAddress)) {
-                            result = false;
-                        }
-                        else {
-                            i += 4;
-                        }
-                    }
-                    if (result) {
-                        for (; i < (long) nameLength; ++i) {
-                            if (UNSAFE.getByte(existingResult.nameAddress + i) != UNSAFE.getByte(nameAddress + i)) {
-                                result = false;
-                                break;
-                            }
-                        }
-                    }
-                    if (result) {
+                    if (((UNSAFE.getLong(existingResult.nameAddress + i) ^ UNSAFE.getLong(nameAddress + i)) << (64 - (nameLength - i) << 3)) == 0) {
                         existingResult.min = (short) Math.min(existingResult.min, number);
```

Comment:
Looks like these cast would now be unnecessary like in lines 69-70

Reply:
Yes, agreed, thanks for spotting. Left over from last version and now this might actually add an additional instruction!